pkg/ast/spl/spl.peg

```diff
@@ -40,6 +40,130 @@ func ConvertCustomDateTimeFormatToEpochMs(dateStr string) (int64, error) {
+                return time.Date(year, mon, day-int(diff), 0, 0, 0, 0, time.Local), nil
+            case utils.TMMonth:
+                return time.Date(year, mon, 1, 0, 0, 0, 0, time.Local), nil
+            case utils.TMQuarter:
+                if mon >= time.October {
+                    mon = time.October
+                } else if mon >= time.July {
+                    mon = time.July
+                } else if mon >= time.April {
+                    mon = time.April
+                } else {
+                    mon = time.January
+                }
+                return time.Date(year, mon, 1, 0, 0, 0, 0, time.Local), nil
+            case utils.TMYear:
+                return time.Date(year, 1, 1, 0, 0, 0, 0, time.Local), nil
+            default:
+                return t, fmt.Errorf("Unsupported time unit for relative timestamp: %v", tunit)
+        }
+    } else {
+        if len(snap) != 2 {
+            return t, fmt.Errorf("Error for special week snap, should be of type w[0-7] got: %v", snap)
+        }
+        weeknum := int(snap[1] - '0')
+        if weeknum == 7 {
+            weeknum = 0
+        }
+        diff := int(week)-weeknum
+        if diff < 0 {
+            diff += 7
+        }
+        return time.Date(year, mon, day, 0, 0, 0, 0, time.Local).AddDate(0, 0, -diff), nil
+    }
+}
+
+
+func applyOffset(num int64, unit utils.TimeUnit, t time.Time) (time.Time, error) {
+
+    durNum := time.Duration(num)
+
+    switch unit {
+        case utils.TMSecond:
+            return t.Add(durNum * time.Second), nil
+        case utils.TMMinute:
+            return t.Add(durNum * time.Minute), nil
+        case utils.TMHour:
+            return t.Add(durNum * time.Hour), nil
+        case utils.TMDay:
+            return t.AddDate(0, 0, int(num)), nil
+        case utils.TMWeek:
+            return t.AddDate(0, 0, 7 * int(num)), nil
+        case utils.TMMonth:
+            return t.AddDate(0, int(num), 0), nil
+        case utils.TMQuarter:
+            return t.AddDate(0, 4*int(num), 0), nil
+        case utils.TMYear:
+            return t.AddDate(int(num), 0, 0), nil
+        default:
+            return t, fmt.Errorf("Unsupported time unit for relative timestamp: %v", unit)
+    }
+}
+
+func calculateRelativeTime(relTimeModifier ast.RelativeTimeModifier, currTime time.Time) (int64, error) {
+    var epoch int64 = 0
+    var err error
+    if relTimeModifier.AbsoluteTime == "now" {
+        epoch = currTime.UnixMilli()
+    } else if relTimeModifier.AbsoluteTime == "1" {
+        return 1, nil
+    } else if relTimeModifier.AbsoluteTime != "" {
+        epoch, err = ConvertCustomDateTimeFormatToEpochMs(relTimeModifier.AbsoluteTime)
+        if err != nil {
+            return 0, err
+        }
+    } else {
+        offsetNum := relTimeModifier.RelativeTimeOffset.Offset
+        offsetTimeUnit := relTimeModifier.RelativeTimeOffset.TimeUnit
+        snap := relTimeModifier.Snap
+        t := currTime
```

Comment:
Why are you renaming `currTime` to `t`? It seems like this only makes the code less clear

Reply:
We are operating on t to apply offset and then whatever we get we would apply snap if present on that time. setting the currTime to offsetApplied currTime would be a bit unclear so just used a variable to represent the currTime.

Reply:
Ok, then maybe naming it something like `finalTime` would be more clear?

Aside from loop indexes, I generally find that 1-letter names are not very clear, and once they're in the codebase it makes searching for them a little more annoying because you have to do a regex search like `[^a-z]t[^a-z]` instead of just `myVariable` or else you'll get a lot of matches in the middle of other words

Reply:
addressed in the latest commit